mautic/mautic
# Complete refactoring of campaign processing

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? | todo
| Related user documentation PR URL |  
| Related developer documentation PR URL | todo
| Issues addressed (#s or URLs) | Happens to fix #4494
| BC breaks? | no
| Deprecations? | lots

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This is a complete refactoring of campaign processing. It allows consuming contacts in batches to prevent the overhead of one by one. This allows actions that support batches of contacts (such as sending to a API based email transport) to consume in batches. 

This also in theory sets up the framework required to send contacts through a campaign multiple times and to jump to other areas of a campaign. 

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Test everything campaigns :-) 

Some immediate test cases that come to mind: 
1. Decision (submits form) at first level of a campaign with actions connected into both active/inactive paths. Actions on active should be executed as soon s the form is submitted. The inactive paths should be executed if the form isn't submitted within the set timeframe (test both interval and date options)
2. When an inactive path of a decision has multiple time frames to determine inactivity, the earliest one is used to determine inactive status then the rest are scheduled. 
3. Same as number 1 but not as the first level of the campaign. 
4. Chained conditions are processed correctly
5. Chained actions and conditions (action -> condition -> action -> condition, etc) are processed correctly
6. A contact is half way through a campaign, we then edit the campaign to add a new action to the beginning, that contact should not go back and execute the newly added action. 
7. Scheduled actions/conditions are processed correctly. 
8. Test that marketing messages still send based on channel priority

#### List deprecations along with the new alternative:
1. To do
2.
